cmd/local-locker.go

```diff
@@ -164,9 +164,19 @@ func (l *localLocker) RLock(ctx context.Context, args dsync.LockArgs) (reply boo
 	if len(args.Resources) != 1 {
 		return false, fmt.Errorf("internal error: localLocker.RLock called with more than one resource")
 	}
+	if args.TimeoutMillis > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(args.TimeoutMillis)*time.Millisecond)
+		defer cancel()
+	}
 
 	l.mutex.Lock()
 	defer l.mutex.Unlock()
+
```

Comment:
It would be even better to turn the mutex into a channel so that we can wait in parallel for acquiring the lock and for the timeout. I am happy to change that as part of the PR if accepted.

Reply:
Doesn't really matter too much, if you do that you would need to spin up a goroutine for every blocked request, that would live until the lock is acquired.

Reply:
My thought was to use something like 
```go 
select {
	case <-ctx.Done():
	    return false, ctx.Err()
	case mutex <- struct{}{}:
	}
}
```

Reply:
How would you "feed" the mutex channel?

Reply:
The `mutex` channel would a field of the `localLocker` struct and is initialized to capacity of 1. This means only one Lock/Rlock invocation can send an item to it. Exiting from the method would remove the item from the channel, allowing pending locks to send a new item.. This way a Lock/RLock invocation could return as soon as the timeout expires.

The main difference is that with the current implementation the method still has to lock the mutex even when the timeout has expired, likely causing more contention.

Note that locking the mutex requires sending an item to the channel, and unlocking it requires reading an item from it.

Reply:
I have tested that already and the raw throughput tanks compared to a mutex. You are welcome to try -`Test_localLocker_RUnlock` should be fine for benchmarking baseline.